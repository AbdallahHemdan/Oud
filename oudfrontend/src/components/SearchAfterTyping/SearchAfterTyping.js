import React, { Component } from 'react'
import axios from "axios";
import { base, prodUrl } from "./../../config/environment"
import SearchCategory from './../SearchCategory/SearchCategory';
import PropTypes from "prop-types";

/**
 *  component to render all the search results for specific search value
 *
 * @author Abdallah Hemdan
 *
 * @component
 *
 */
class SearchAfterTyping extends Component {
  /**
   * @constructor
   *
   * @param {object} props - get props from higher components (Search)
   * @param {Array<Object>} tracks - List of tracks for search value
   * @param {Array<Object>} albums - List of albums for search value
   * @param {Array<Object>} artists - List of artists for search value
   * @param {Array<Object>} users - List of users for search value
   * @param {Array<Object>} playlists - List of playlists for search value
   * @param {Number} limit - The maximum number of categories to get.
   * @param {Number} offset - The index of the first categories to get.
   * @param {Number} total - The total number of categories available to get.
   * @param {String} search - The search value that user type
   * @param {Boolean} isLoading - variable to check if the data is loaded or not
   */
  constructor(props) {
    super(props)
    this.state = {
      /**
       * 
       * List of tracks for search value
       * 
       * @@type {Array<Object>}
       */
      tracks: [],
      /**
       *
       * List of albums for search value
       *
       * @@type {Array<Object>}
       */
      albums: [],
      /**
       *
       * List of artists for search value
       *
       * @@type {Array<Object>}
       */
      artists: [],
      /**
       *
       * List of users for search value
       *
       * @@type {Array<Object>}
       */
      users: [],
      /**
       *
       * List of playlists for search value
       *
       * @@type {Array<Object>}
       */
      playlists: [],
      /**
       * The maximum number of categories to get.
       *
       * @type {number}
       */
      limit: 0,

      /**
       * The index of the first categories to get.
       *
       * @type {number}
       */
      offset: 0,
      /**
       * The total number of categories available to get.
       *
       * @type {number}
      */
      total: 0,
      /**
       * the value of the search that the user has type
       * @type {String}
       */
      search: "",
      /**
       * Check if the data loaded from the backend or not
       * @type {Boolean}
       */
      isLoading: true
    }
  }
  /**
   * Function to handle store search value that user type in the state..
   *
   * @function
   *
   * @returns {void}
   */
  handleStoringData = ({ tracks, albums, artists, users, playlists, limit, offset, total }) => {
    this.setState({ tracks, albums, artists, users, playlists, limit, offset, total, isLoading: false });
  }

  /**
   * Function to fetch search items of specific search value that user type
   * 
   * @function
   *
   * @returns {void}
  */
  componentDidUpdate() {
    if (this.props.canSend) {
      const fetchAllUrl = (base === prodUrl) ? `${base}/search/?q=${this.props.search}` : `${base}/search`;
      if (this.state.search !== this.props.search) {
        axios.get(fetchAllUrl)
          .then((result) => {
            this.handleStoringData(result.data);
          }).catch((err) => {
            console.log(err)
          });
        this.setState({ search: this.props.search });
      }
    }
  }

  /**
     * @function
     * @name render
     * @description Render all the Search fetched items for specific search value that user type
     *
     * @returns {JSX} Component for Home
    */
  render() {
    return (
      <React.Fragment>
        {
          (this.state.isLoading) ?
            (<h1 data-testid="loading">Loading..</h1>) :
            (
              <div
                data-testid="search-results"
              >
                {
                  (this.state.tracks.tracks.length) ?
                    <SearchCategory
                      search={this.state.search}
                      propsSearch={this.props.search}
                      name={"Tracks"}
                      items={this.state.tracks}
                      type="albums"
                      data-testid="search-tracks"
                    /> : null
                }

                {
                  (this.state.artists.artists.length) ?
                    <SearchCategory
                      search={this.state.search}
                      propsSearch={this.props.search}
                      name={"Artists"}
                      items={this.state.artists}
                      type="artist"
                      data-testid="search-artists"
                    /> : null
                }

                {
                  (this.state.albums.albums.length) ?
                    <SearchCategory
                      search={this.state.search}
                      propsSearch={this.props.search}
                      name={"Albums"}
                      items={this.state.albums}
                      type="albums"
                      data-testid="search-albums"
                    /> : null
                }
                {
                  (this.state.playlists.playlists.length) ?
                    < SearchCategory
                      search={this.state.search}
                      propsSearch={this.props.search}
                      name={"Playlists"}
                      items={this.state.playlists}
                      type="playlist"
                      data-testid="search-playlists"
                    /> : null
                }
                {
                  (this.state.users.users.length) ?
                    <SearchCategory
                      search={this.state.search}
                      propsSearch={this.props.search}
                      name={"Users"}
                      items={this.state.users}
                      type="profile"
                      data-testid="search-users"
                    /> : null
                }
              </div>
            )
        }
      </React.Fragment>
    )
  }
}

SearchAfterTyping.propTypes = {
  search: PropTypes.string,
  canSend: PropTypes.bool
}

export default SearchAfterTyping
