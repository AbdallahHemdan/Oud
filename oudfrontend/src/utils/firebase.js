import firebase from "firebase";

const firebasConfig = {
  apiKey: "AIzaSyBZVUambAFEB-ywzAZTAPP-tSkbYXSEWc4",
  authDomain: "oud-notifications.firebaseapp.com",
  databaseURL: "https://oud-notifications.firebaseio.com",
  projectId: "oud-notifications",
  storageBucket: "oud-notifications.appspot.com",
  messagingSenderId: "681943816425",
  appId: "1:681943816425:web:9149a8834b2ff0f4623878"
};
// Initialize Firebase
firebase.initializeApp(firebasConfig);

export default firebase;
