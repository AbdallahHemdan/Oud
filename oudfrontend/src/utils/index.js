import axios from 'axios'
import { base } from '../config/environment'
const findByTestAttr = (component, val) => {
    return component.find(`[data-testid="${val}"]`);
}

function addToLikedSongs(id) {
    axios.put(`${base}/me/tracks/${id}`)
        .then(function (response) {
            console.log(response);

        })
        .catch(function (error) {
            console.log(error);
        });
}
function removeLikedSong(id) {
    axios.delete(`${base}/me/tracks/${id}`)
        .then(function (response) {
            console.log(response);
        })
        .catch(function (error) {
            console.log(error);
        });
}
export { findByTestAttr, addToLikedSongs, removeLikedSong }
