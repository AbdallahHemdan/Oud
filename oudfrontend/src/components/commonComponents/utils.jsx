import axios from "axios";
import { base } from "../../config/environment"
import {config} from "../../utils/auth"
/**
 * @file this file contains three functions used in Playlist, Album, LikedSongs 
 **/
export function addSong(playlistId, track){
  axios
  .post(`${base}/playlists/${playlistId}/${track}`, config)
  .then((response) => {
  })
  .catch((error) => {
  });
}