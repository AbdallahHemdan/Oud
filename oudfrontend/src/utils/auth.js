import jwtDecode from "jwt-decode";
import { getRequest } from "./requester";
import { base } from "./../config/environment";

function getToken() {
  return localStorage.getItem("accessToken");
}

const config = {
  headers: { authorization: `Bearer ${getToken()}` }
};

function Auth() {
  try {
    if (!getToken()) return null;
    let decoded = jwtDecode(getToken());
    console.log(decoded.id);
    return decoded.id;
  } catch {
    return null;
  }
}
function isLoggedIn() {
  return getToken() ? true : false;
}

function isArtist() {
  return getRequest(`${base}/me`)
    .then(response => {
      return response.data.role === "artist";
    })
    .catch(error => {
      console.log(error);
    });
}
const _config = config;
export { _config as config };
export const token = getToken();
const _Auth = Auth;
export { _Auth as Auth };
const _isLoggedIn = isLoggedIn;
export { _isLoggedIn as isLoggedIn };
const _isArtist = isArtist;
export { _isArtist as isArtist };
