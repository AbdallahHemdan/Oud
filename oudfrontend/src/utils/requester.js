import axios from "axios";
import { isLoggedIn } from "./../utils/auth";
const config = {
  headers: {
    authorization: `Bearer eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpZCI6IjVlOTA3ZGIwYTA2NDVmNDU4MTYwNzYwNiIsImlhdCI6MTU5MTYyMTQxOSwiZXhwIjoxNTk0MjEzNDE5fQ.fj3N3Pc89Pf_xlt7fGmXw1SINTecUB4-y3pihAAPjC8`,
    "Access-Control-Allow-Origin": "*"
  }
};
const getRequest = endpoint => {
  return axios.get(endpoint, config);
};
/**
 * Axios DELETE request
 * @function
 * @param {string} endpoint the required endpoint to implement an action
 * @returns {object}
 */
const deleteRequest = endpoint => {
  return axios.delete(endpoint, config);
};
/**
 * Axios PUT request
 * @function
 * @param {string} endpoint the required endpoint to implement an action
 * @param {object} body the request body
 * @returns {object}
 */
const putRequest = (endpoint, body = {}) => {
  return axios.put(endpoint, body, config);
};
/**
 * Axios POST request
 * @function
 * @param {string} endpoint the required endpoint to implement an action
 * @param {object} body the request body
 * @returns {object}
 */
const postRequest = (endpoint, body = {}) => {
  return axios.post(endpoint, body, config);
};
/**
 * Axios PATCH request
 * @function
 * @param {string} endpoint the required endpoint to implement an action
 * @param {object} body the request body
 * @returns {object}
 */
const patchRequest = (endpoint, body = {}) => {
  return axios.patch(endpoint, body, config);
};

// const isArtist = () => {
//   if (isLoggedIn()) {
//   }
// };
export { getRequest, deleteRequest, putRequest, postRequest, patchRequest };
