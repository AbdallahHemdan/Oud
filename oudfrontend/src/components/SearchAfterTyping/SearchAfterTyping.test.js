import React from "react";
import { shallow, mount } from 'enzyme';
import renderer from 'react-test-renderer'
import { findByTestAttr } from "./../../utils/index"
import SearchAfterTyping from './SearchAfterTyping';
import checkPropTypes from 'check-prop-types'

const setup = (props = {}) => {
  const component = shallow(
    <SearchAfterTyping {...props} />
  )
  return component;
}
const item = {
  search: "Amr",
  canSend: true,
}

describe('Search After Typing component', () => {
  let component;
  beforeEach(() => {
    const props = {
      item
    }
    component = setup(props);
  })
  it('Should render loading in case of loading data', () => {
    component.setState({ isLoading: true })
    const wrapper = findByTestAttr(component, "loading");
    expect(component.state().isLoading).toBeTruthy();
    expect(wrapper.length).toBe(1);
  });


  describe('snapshot test', () => {
    it('renders component correctly', () => {
      const fullProps = { search: "Hemdan", canSend: true };
      const tree = renderer
        .create(<SearchAfterTyping {...fullProps} />)
        .toJSON();
      expect(tree).toMatchSnapshot();
    });
  });

  describe('testing prop types', () => {
    it('should pass true props', () => {
      const result = checkPropTypes(SearchAfterTyping.propTypes, { ...item }, 'prop', SearchAfterTyping.name);
      expect(result).toBeUndefined();
    });
    it('should pass false props', () => {
      const result = checkPropTypes(SearchAfterTyping.propTypes, { search: 0 }, 'prop', SearchAfterTyping.name);
      expect(result).toBeDefined();
    });
    it('should pass false props', () => {
      const result = checkPropTypes(SearchAfterTyping.propTypes, { search: true }, 'prop', SearchAfterTyping.name);
      expect(result).toBeDefined();
    });
    it('should pass false props', () => {
      const result = checkPropTypes(SearchAfterTyping.propTypes, { search: jest.fn }, 'prop', SearchAfterTyping.name);
      expect(result).toBeDefined();
    });
    it('should pass false props', () => {
      const result = checkPropTypes(SearchAfterTyping.propTypes, { search: [15] }, 'prop', SearchAfterTyping.name);
      expect(result).toBeDefined();
    });
    it('should pass false props', () => {
      const result = checkPropTypes(SearchAfterTyping.propTypes, { search: {} }, 'prop', SearchAfterTyping.name);
      expect(result).toBeDefined();
    });
    it('should pass false props', () => {
      const result = checkPropTypes(SearchAfterTyping.propTypes, { canSend: 0 }, 'prop', SearchAfterTyping.name);
      expect(result).toBeDefined();
    });
    it('should pass false props', () => {
      const result = checkPropTypes(SearchAfterTyping.propTypes, { canSend: "" }, 'prop', SearchAfterTyping.name);
      expect(result).toBeDefined();
    });
    it('should pass false props', () => {
      const result = checkPropTypes(SearchAfterTyping.propTypes, { canSend: jest.fn }, 'prop', SearchAfterTyping.name);
      expect(result).toBeDefined();
    });
    it('should pass false props', () => {
      const result = checkPropTypes(SearchAfterTyping.propTypes, { canSend: [15] }, 'prop', SearchAfterTyping.name);
      expect(result).toBeDefined();
    });
    it('should pass false props', () => {
      const result = checkPropTypes(SearchAfterTyping.propTypes, { canSend: {} }, 'prop', SearchAfterTyping.name);
      expect(result).toBeDefined();
    });
  })
});
