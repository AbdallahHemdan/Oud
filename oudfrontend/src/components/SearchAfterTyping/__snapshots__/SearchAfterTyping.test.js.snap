// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Search After Typing component snapshot test renders component correctly 1`] = `
<h1
  data-testid="loading"
>
  Loading..
</h1>
`;
